File: mini/src/lib/dose-calculator.ts
import type { Cartridge } from '../types'

export interface DoseStatus {
  totalDoses: number
  usedDoses: number
  remainingDoses: number
  remainingDays: number
  currentDailyDoses: number
  estimatedEndDate: string  // YYYY-MM-DD
}

/**
 * Calculate the number of days between two YYYY-MM-DD date strings.
 * Same day returns 0. Uses UTC to avoid timezone issues.
 */
function daysBetween(dateA: string, dateB: string): number {
  const a = new Date(dateA + 'T00:00:00Z')
  const b = new Date(dateB + 'T00:00:00Z')
  return Math.round((b.getTime() - a.getTime()) / (1000 * 60 * 60 * 24))
}

/**
 * Add a number of days to a YYYY-MM-DD date string and return a new YYYY-MM-DD string.
 */
function addDays(date: string, days: number): string {
  const d = new Date(date + 'T00:00:00Z')
  d.setUTCDate(d.getUTCDate() + days)
  return d.toISOString().slice(0, 10)
}

/**
 * Calculate the current dose status for a cartridge on a given day.
 *
 * Logic:
 * 1. Find the latest manual adjustment on or before `today` to use as baseline.
 *    If none, use (startDate, totalDoses) as baseline.
 * 2. From baseDate to today, sum used doses by iterating through dosageChange segments.
 * 3. remainingDoses = max(0, baseDoses - usedSinceBase)
 * 4. remainingDays = floor(remainingDoses / currentDailyDoses)
 * 5. estimatedEndDate = today + remainingDays
 */
export function calculateDoseStatus(cartridge: Cartridge, today: string): DoseStatus {
  // Step 1: Determine baseline (baseDate, baseDoses)
  const validAdjustments = cartridge.manualAdjustments
    .filter(a => a.date <= today)
    .sort((a, b) => a.date.localeCompare(b.date))

  let baseDate: string
  let baseDoses: number

  if (validAdjustments.length > 0) {
    const latest = validAdjustments[validAdjustments.length - 1]
    baseDate = latest.date
    baseDoses = latest.remainingDoses
  } else {
    baseDate = cartridge.startDate
    baseDoses = cartridge.totalDoses
  }

  // Step 2: Calculate used doses from baseDate to today
  const sortedChanges = [...cartridge.dosageChanges]
    .sort((a, b) => a.date.localeCompare(b.date))

  // Find the active daily dose rate at baseDate (latest change with date <= baseDate)
  let activeDailyDoses = 0
  for (const change of sortedChanges) {
    if (change.date <= baseDate) {
      activeDailyDoses = change.dailyDoses
    }
  }

  // Iterate through dosage changes that fall strictly after baseDate and on or before today,
  // building segments and summing used doses.
  let usedSinceBase = 0
  let segmentStart = baseDate
  let currentRate = activeDailyDoses

  for (const change of sortedChanges) {
    if (change.date > baseDate && change.date <= today) {
      const days = daysBetween(segmentStart, change.date)
      usedSinceBase += days * currentRate
      segmentStart = change.date
      currentRate = change.dailyDoses
    }
  }

  // Final segment from last segment start to today
  const finalDays = daysBetween(segmentStart, today)
  usedSinceBase += finalDays * currentRate

  // Step 3: Remaining doses (never below 0)
  const remainingDoses = Math.max(0, baseDoses - usedSinceBase)

  // Current daily doses = the rate active at today (latest change with date <= today)
  let currentDailyDoses = 0
  for (const change of sortedChanges) {
    if (change.date <= today) {
      currentDailyDoses = change.dailyDoses
    }
  }

  // Step 4: Remaining days
  const remainingDays = currentDailyDoses > 0
    ? Math.floor(remainingDoses / currentDailyDoses)
    : 0

  // Step 5: Estimated end date
  const estimatedEndDate = addDays(today, remainingDays)

  // usedDoses is relative to the total cartridge capacity
  const usedDoses = cartridge.totalDoses - remainingDoses

  return {
    totalDoses: cartridge.totalDoses,
    usedDoses,
    remainingDoses,
    remainingDays,
    currentDailyDoses,
    estimatedEndDate,
  }
}

File: ios/MedicationReminder/MedicationReminder/DoseCalculator.swift
import Foundation

struct DoseStatus: Equatable {
    var totalDoses: Int
    var usedDoses: Int
    var remainingDoses: Int
    var remainingDays: Int
    var currentDailyDoses: Int
    var estimatedEndDate: String
}

private let utcCalendar: Calendar = {
    var calendar = Calendar(identifier: .gregorian)
    calendar.timeZone = TimeZone(secondsFromGMT: 0)!
    return calendar
}()

private func date(from value: String) -> Date {
    var components = DateComponents()
    components.calendar = utcCalendar
    components.timeZone = TimeZone(secondsFromGMT: 0)
    components.year = Int(value.prefix(4))
    components.month = Int(value.dropFirst(5).prefix(2))
    components.day = Int(value.dropFirst(8).prefix(2))
    return utcCalendar.date(from: components) ?? Date(timeIntervalSince1970: 0)
}

private func daysBetween(_ start: String, _ end: String) -> Int {
    let components = utcCalendar.dateComponents([.day], from: date(from: start), to: date(from: end))
    return components.day ?? 0
}

private func addDays(_ value: String, _ days: Int) -> String {
    let next = utcCalendar.date(byAdding: .day, value: days, to: date(from: value)) ?? date(from: value)
    return dateString(from: next)
}

func dateString(from value: Date) -> String {
    let components = utcCalendar.dateComponents([.year, .month, .day], from: value)
    return String(format: "%04d-%02d-%02d", components.year ?? 1970, components.month ?? 1, components.day ?? 1)
}

func todayString() -> String {
    dateString(from: Date())
}

func timestampString() -> String {
    let formatter = ISO8601DateFormatter()
    formatter.formatOptions = [.withInternetDateTime]
    return String(formatter.string(from: Date()).prefix(19))
}

func calculateDoseStatus(cartridge: Cartridge, today: String) -> DoseStatus {
    let validAdjustments = cartridge.manualAdjustments
        .filter { $0.date <= today }
        .sorted { $0.date < $1.date }

    let baseDate: String
    let baseDoses: Int

    if let latest = validAdjustments.last {
        baseDate = latest.date
        baseDoses = latest.remainingDoses
    } else {
        baseDate = cartridge.startDate
        baseDoses = cartridge.totalDoses
    }

    let sortedChanges = cartridge.dosageChanges.sorted { $0.date < $1.date }

    var activeDailyDoses = 0
    for change in sortedChanges where change.date <= baseDate {
        activeDailyDoses = change.dailyDoses
    }

    var usedSinceBase = 0
    var segmentStart = baseDate
    var currentRate = activeDailyDoses

    for change in sortedChanges where change.date > baseDate && change.date <= today {
        usedSinceBase += daysBetween(segmentStart, change.date) * currentRate
        segmentStart = change.date
        currentRate = change.dailyDoses
    }

    usedSinceBase += daysBetween(segmentStart, today) * currentRate

    let remainingDoses = max(0, baseDoses - usedSinceBase)

    var currentDailyDoses = 0
    for change in sortedChanges where change.date <= today {
        currentDailyDoses = change.dailyDoses
    }

    let remainingDays = currentDailyDoses > 0 ? remainingDoses / currentDailyDoses : 0
    let usedDoses = cartridge.totalDoses - remainingDoses

    return DoseStatus(
        totalDoses: cartridge.totalDoses,
        usedDoses: usedDoses,
        remainingDoses: remainingDoses,
        remainingDays: remainingDays,
        currentDailyDoses: currentDailyDoses,
        estimatedEndDate: addDays(today, remainingDays)
    )
}
